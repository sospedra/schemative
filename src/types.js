import { assign, reduce, noop } from 'lodash'
import PropTypes from 'prop-types'

const EXPOSED_PROPERTIES = ['value']

/**
 * Trigger a warning when assigning a value which does not
 * fit with the provided scalar.
 * Return default value.
 *
 * @param  {Any} candidate
 * @param  {Any} fallback
 * @param  {String} scalar
 * @return {Any}
 */
const evaluateValueType = (candidate, fallback, scalar) => {
  if (scalar !== 'any' && typeof candidate !== typeof fallback) {
    console.warn(
      'Evalute Schemative type expected',
      typeof fallback, 'but receive', typeof candidate,
      'when', candidate, 'value was set.',
      'Default value is returned instead.'
    )

    return fallback
  }

  return candidate
}

/**
 * Given a value, type base and default value creates an Schemative type.
 * Create the object with the type definition.
 * Pick all the properties that goes to the prototype.
 * Add the properties to an object containing the type exposed values.
 *
 * @param  {Any} value        - Selected value
 * @param  {Object} base      - Contains basic type's data
 *   @param  {String} scalar    - Type name
 *   @param  {Func} propTypes   - React PropTypes equivalent
 * @param  {Any} defaultValue - Default type value
 * @return {Object}           - Type
 */
const createType = (value, base, defaultValue) => {
  const type = assign({}, base, {
    isRequired: !base.__required__ &&
      createType(value, createTypeBase(base, true), defaultValue)
  })
  const properties = reduce(type, (memo, value, property) => {
    if (EXPOSED_PROPERTIES.includes(property)) return memo

    return assign({}, memo, {
      [property]: { get: () => value }
    })
  }, {})

  return Object.defineProperties({
    value: evaluateValueType(value, defaultValue, base.scalar)
  }, properties)
}

const createTypeBase = (base, isRequired) => {
  const propType = PropTypes[base.scalar]

  return {
    ...base,
    __required__: !!isRequired,
    propType: isRequired ? propType.isRequired : propType,
    scalar: base.scalar
  }
}

const createTypeWithDefault = (scalar, defaultValue) => {
  const base = createType(defaultValue, createTypeBase({ scalar }), defaultValue)
  const type = (value) => createType(value, base, defaultValue)

  Object.getOwnPropertyNames(base).forEach((property) => {
    if (EXPOSED_PROPERTIES.includes(property)) {
      type[property] = base[property]
    } else {
      Object.defineProperty(type, property, {
        get: () => base[property]
      })
    }
  })

  return type
}

export const any = createTypeWithDefault('any', true)
export const array = createTypeWithDefault('array', [])
export const bool = createTypeWithDefault('bool', true)
export const element = createTypeWithDefault('element', noop)
export const func = createTypeWithDefault('func', noop)
export const number = createTypeWithDefault('number', 0)
export const object = createTypeWithDefault('object', {})
export const shape = createTypeWithDefault('shape', {})
export const string = createTypeWithDefault('string', '')
export const symbol = createTypeWithDefault('symbol', Symbol(''))
