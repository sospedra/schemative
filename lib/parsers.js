'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDefault = exports.createPropTypes = exports.recursiveValues = exports.selectRecursiveStrategy = exports.recursiveArray = exports.recursiveObject = exports.getPropType = undefined;

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var getPropType = exports.getPropType = function getPropType(prop, attr) {
  return prop.iterable ? recursiveValues(prop, attr) : prop[attr];
};

var recursiveObject = exports.recursiveObject = function recursiveObject(values, attr) {
  return _lodash2.default.reduce(values, function (memo, prop, key) {
    return Object.assign(_defineProperty({}, key, getPropType(prop, attr)), memo);
  }, {});
};

var recursiveArray = exports.recursiveArray = function recursiveArray(values, attr) {
  return _lodash2.default.map(values, function (prop) {
    return getPropType(prop, attr);
  });
};

var selectRecursiveStrategy = exports.selectRecursiveStrategy = function selectRecursiveStrategy(values, attr) {
  return _lodash2.default.isArray(values) ? recursiveArray(values, attr) : recursiveObject(values, attr);
};

var recursiveValues = exports.recursiveValues = function recursiveValues(_ref, attr) {
  var type = _ref.type,
      values = _ref.values;

  var exec = attr === 'type' ? type : _lodash2.default.identity;
  return exec(selectRecursiveStrategy(values, attr));
};

var createPropTypes = exports.createPropTypes = function createPropTypes(definition) {
  return selectRecursiveStrategy(definition, 'type');
};

var createDefault = exports.createDefault = function createDefault(definition) {
  return selectRecursiveStrategy(definition, 'default');
};
//# sourceMappingURL=parsers.js.map