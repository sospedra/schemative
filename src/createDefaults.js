import tsil from 'tsil'

export default (nodes) => {
  const defaultValues = tsil.modify(nodes, (node) => {
    return node.value
  })

  return tsil.deflatten(defaultValues)
}
