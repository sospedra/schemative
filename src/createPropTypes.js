import tsil from 'tsil'

export default (nodes) => {
  const propTypes = tsil.modify(nodes, (node) => {
    return node.propType
  })

  return tsil.deflatten(propTypes)
}
