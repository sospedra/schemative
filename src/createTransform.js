export default (nodes) => {
  // console.log('createTransform')
}
