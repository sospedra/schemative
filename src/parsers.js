import _ from 'lodash'

export const getPropType = (prop, attr) => {
  return prop.iterable ? recursiveValues(prop, attr) : prop[attr]
}

export const recursiveObject = (values, attr) => {
  return _.reduce(values, (memo, prop, key) => Object.assign({
    [key]: getPropType(prop, attr)
  }, memo), {})
}

export const recursiveArray = (values, attr) => {
  return _.map(values, (prop) => {
    return getPropType(prop, attr)
  })
}

export const selectRecursiveStrategy = (values, attr) => {
  return _.isArray(values)
    ? recursiveArray(values, attr)
    : recursiveObject(values, attr)
}

export const recursiveValues = ({ type, values }, attr) => {
  const exec = attr === 'type' ? type : _.identity
  return exec(selectRecursiveStrategy(values, attr))
}

export const createPropTypes = (definition) => {
  return recursiveObject(definition, 'type')
}

export const createDefault = (definition) => {
  return recursiveObject(definition, 'default')
}

export const filterByKeys = (definition, candidate) => {
  const keys = _.intersection(_.keys(definition), _.keys(candidate))
  return _.pick(candidate, keys)
}

export const transform = (definition, candidate, mutators) => Object.assign(
  filterByKeys(definition, candidate),
  _.reduce(mutators, (memo, value, key) => Object.assign({
    [key]: _.isFunction(value) ? value(candidate) : value,
    memo
  }), {})
)
