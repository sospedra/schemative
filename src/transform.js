import _ from 'lodash'

export const filterByKeys = (defsKeys, candidate) => {
  const keys = _.intersection(defsKeys, _.keys(candidate))
  return _.pick(candidate, keys)
}

export const createTransform = (definitions) => {
  const defsKeys = _.keys(definitions)
  let protoMutators = {}

  const transform = (candidate, mutators = protoMutators) => {
    return Object.assign(
      filterByKeys(defsKeys, candidate),
      _.reduce(mutators, (memo, value, key) => Object.assign({
        [key]: _.isFunction(value) ? value(candidate, definitions) : value
      }, memo), {})
    )
  }

  Object.defineProperty(transform, 'mutators', {
    get: () => protoMutators,
    set: (mutators) => {
      protoMutators = mutators
    }
  })

  return transform
}
