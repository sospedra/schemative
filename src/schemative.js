import { PropTypes } from 'react'
import _ from 'lodash/util'
import { createPropTypes, createDefault } from './parsers'
import { createTransform } from './transform'

const boundIsRequire = (prototype) => {
  return Object.defineProperty(prototype, 'isRequired', {
    value: Object.assign({}, prototype, {
      type: prototype.type.isRequired
    })
  })
}

const createType = (type, def) => boundIsRequire({
  type: PropTypes[type],
  default: def,
  iterable: false
})

const createFunctionType = (type) => (values) => boundIsRequire({
  type: PropTypes[type],
  values: values,
  iterable: true
})

export const any = createType('any', _.stubTrue())
export const array = createType('array', _.stubArray())
export const bool = createType('bool', _.stubTrue())
export const element = createType('element', _.stubFalse())
export const func = createType('func', _.noop())
export const instanceOf = createType('instanceOf', _.stubFalse())
export const number = createType('number', 0)
export const node = createType('node', _.stubFalse())
export const object = createType('object', _.stubObject())
export const string = createType('string', _.stubString())
export const symbol = createType('symbol', Symbol)
export const arrayOf = createFunctionType('arrayOf')
export const objectOf = createFunctionType('objectOf')
export const oneOf = createFunctionType('oneOf')
export const oneOfType = createFunctionType('oneOfType')
export const shape = createFunctionType('shape')

export const createSchema = (definition) => ({
  PropTypes: createPropTypes(definition),
  Default: createDefault(definition),
  transform: createTransform(definition)
})
