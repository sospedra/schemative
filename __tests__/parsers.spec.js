/* global describe, it, expect */
import * as parsers from '../src/parsers'
import * as mocks from './mocks'

describe('Suite parsers', () => {
  const {
    getPropType,
    recursiveObject,
    recursiveArray,
    selectRecursiveStrategy,
    recursiveValues,
    createPropTypes,
    createDefault,
    filterByKeys,
    transform
  } = parsers

  it('should return the defined value or a recursion', () => {
    expect(getPropType(mocks.defs.simple.name, 'type')).toBe(mocks.defs.simple.name.type)
    expect(getPropType(mocks.defs.simple.name, 'default')).toBe(mocks.defs.simple.name.default)
    expect(getPropType(mocks.defs.object, 'type')).toMatchSnapshot()
  })

  it('should create an schema given an object', () => {
    expect(recursiveObject(mocks.defs.object)).toMatchSnapshot()
  })

  it('should create an schema given an array', () => {
    expect(recursiveArray(mocks.defs.array)).toMatchSnapshot()
  })

  it('should select an strategy for recursivity', () => {
    expect(selectRecursiveStrategy(mocks.defs.array)).toMatchSnapshot()
    expect(selectRecursiveStrategy(mocks.defs.object)).toMatchSnapshot()
  })

  it('should create the schema prop types', () => {
    expect(createPropTypes(mocks.defs.schemativePropTypes)).toMatchSnapshot()
  })

  it('should create the schema prop types with isRequired', () => {
    expect(createPropTypes(mocks.defs.isRequiredPropTypes)).toMatchSnapshot()
  })

  it('should create the schema default values', () => {
    expect(createDefault(mocks.defs.schemativePropTypes)).toMatchSnapshot()
  })

  it('should filter by keys two objects', () => {
    expect(filterByKeys(mocks.defs.simple, mocks.defs.complex)).toEqual(mocks.defs.simple)
  })

  it('should transform a definition into a fulfilled object', () => {
    expect(transform(mocks.defs.simple, mocks.fulfilled)).toEqual(mocks.fulfilled)
  })

  it('should transform a definition using mutators', () => {
    expect(transform(mocks.defs.simple, mocks.fulfilled, mocks.mutators)).toMatchSnapshot()
  })
})
