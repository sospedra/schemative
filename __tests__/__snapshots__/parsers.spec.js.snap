exports[`Suite parsers should create an schema given an array 1`] = `
Array [
  undefined,
  undefined,
  undefined,
]
`;

exports[`Suite parsers should create an schema given an object 1`] = `
Object {
  "iterable": undefined,
  "type": undefined,
  "values": undefined,
}
`;

exports[`Suite parsers should create the schema default values 1`] = `
Object {
  "arr": Array [
    -1,
    "",
  ],
  "dispatch": undefined,
  "func": undefined,
  "shape": Object {
    "deep": Object {
      "nested": "",
    },
    "id": -1,
    "object": Object {},
  },
  "value": -1,
}
`;

exports[`Suite parsers should create the schema prop types 1`] = `
Object {
  "arr": [Function],
  "dispatch": [Function],
  "func": [Function],
  "shape": [Function],
  "value": [Function],
}
`;

exports[`Suite parsers should create the schema prop types with isRequired 1`] = `
Object {
  "geo": [Function],
  "req": [Function],
}
`;

exports[`Suite parsers should return the defined value or a recursion 1`] = `[Function]`;

exports[`Suite parsers should select an strategy for recursivity 1`] = `
Object {
  "iterable": undefined,
  "type": undefined,
  "values": undefined,
}
`;

exports[`Suite parsers should select an strategy for recursivity 2`] = `
Object {
  "iterable": undefined,
  "type": undefined,
  "values": undefined,
}
`;

exports[`Suite parsers should transform a definition using mutators 1`] = `
Object {
  "memo": Object {
    "extra": "attribute",
    "memo": Object {},
  },
  "name": "Ada Lovelace",
  "year": 1852,
}
`;
